### PR TITLE
Extending Governance Levels

### PR DESCRIPTION
**Pull request by robtuley**
Validate and extend the existing "Governance Levels" initial pattern. Proposed new name ideas in PR discussion. 

[Direct link to updated `governance-levels.md`](https://github.com/Flutter-Tech/InnerSourcePatterns/blob/governance-levels/patterns/1-initial/governance-levels.md) 

Changed:

- [x] Add Patlet
- [x] Refactor Context/Problem into more standard Problem/Story
- [x] Re-instate Context
- [x] Review and extend Forces
- [x] Tighten up Solution
- [x] Extend Resulting Context
- [x] Add Known Instance

### CODE REVIEW

**Comment by lenucksi** [comment]
Nice, I like this point. Addition from my side: The academic body of knowledge does qualify both the modes of Project A and Project B as very valid common variations of occurrence of working InnerSource. 

The topic identified here is governance of projects. This is similar to open source where you can have modes A and B and the same resulting frustration as licenses do not imply specific governance. Often this is implied not made explicit. Open Source has the efforts of the SFOSC here trying to provide means for that.

I would argue we should not be prescriptive here but enabling and clarifying, enabling people to make informed decisions. Hence I suggest: "Make governance explicit and provide proven governance models for your InnerSource program" as a goal for the pattern. 
This could work by stating the the governance in the README and possibly provide some approaches to select from in the InnerSource docs for the company.

**Comment by spier** [comment]
This sounds like a well thought out plan to improve this @robtuley. Nice!

Please let us know how we can best help you.

As part of this we might be able to level-up the pattern to level 2 (Structured).

If we end up doing that, some more points to consider:

* **name of the pattern** - re-evaluate if the name describes the pattern will, given that level 2 patterns get published in our online book and changing the name after that is a bit more trouble
* **visualization** -  could any visualization/chart/etc support this pattern?

**Comment by spier** [comment]
@robtuley would you like me to start reviewing the PR already? Or shall I wait until you mark it as "ready for review"? Whatever works best for you :)

**Comment by robtuley** [comment]
@spier intention is to finish it and mark it ready for review at the weekend. Last weeks just went a bit busy at work so not much space for anything else but this isn't forgotten.

**Comment by robtuley** [comment]
@spier -- updated to ready to review as I think this is in good shape now. 

**Naming**

I think a better name for this pattern is one of:

* Common Language
* Universal Language
* Standard Language 
* Menu of Models

I get where "Transparent Governance Levels" is coming from but it doesn't resonant with me personally. I think it focuses the reader's attention on governance processes rather that the root problem of ambiguous language use between people/teams. For me it becomes easier to govern InnerSource practices however you wish if you have a _Common Language_ to more precisely describe it. But I do see that if you have _Transparent Governance Levels_ in place then that should and will form a common language.

However, from @lenucksi [comment earlier](#issuecomment-921626939) -- "The topic identified here is governance of projects" -- I suspect we might not yet be aligned on this point..? 

**Visualisation**

Struggling on this topic -- happy to draw something up but can't think of something that would add value. Any suggestions?

**Comment by spier** [comment]
Thank you for your work on this pattern @robtuley. I am sure there are some great improvements in there already!

I am to review this in more detail next week.
Will also reach out to the community in Slack to see if anybody else has input on this topic.

**Comment by spier** [comment]
Sorry @robtuley, got swamped with work, and did not get to this yet.
But I did not forget about it :)

**Comment by robtuley** [comment]
Know the feeling @spier *chuckle*. No problem at all, there's no particular rush and I appreciate the update.

File: patterns/1-initial/governance-levels.md
Comment (MaineC) [inline patterns/1-initial/governance-levels.md:9]: I think the part about "the term is not precise enough" misses what happens (at least where I have seen this pattern as helpful). Careful looooong explanation coming:


In Scrum, the common saying is that "either you do Scrum fully embracing all aspects of it, or you are not allowed to claim you are doing Scrum".

The thing I like about InnerSource is that the goal is different: Instead of having InnerSource projects/ teams go through a list of checkboxes and being compliant only if all checkboxes are ticked, we provide patterns to fix specific problems the organisation has. The trouble is that a lot of people have now gone through the "if you do not do all of it, you are not doing X right" training. As a result people expect a world that is either black or white: Either projects are InnerSource compliant and operate exactly as I expect them to operate, or they are not. It is tricky to teach that there is a spectrum of practices - and each place on that spectrum comes with it's own costs (in terms of shared communication) and advantages (in terms of collaboration). (The Maturity Model pattern helps to understand, which aspects are touched by InnerSource and likely should be linked here.)

Also sometimes there are corporate constraints to how far you can take InnerSource:

In some companies it may be possible to become a Trusted committer across line management chains. In others this may not be possible.

Now the funny thing is, that the same happens in Open Source (and always have been) - except the "general public" is only just now starting to understand that: https://www.sfscon.it/talks/the-economics-of-vendor-neutrality-and-vendor-domination/

My personal hope with this pattern: Teach your average engineer that there are multiple levels of InnerSource shared governance - and make them recognize that same pattern in Open Source projects:

There is a baseline (using an open source license that guarantees the four freedoms of use, study, share, improve) - but beyond that, open source projects are governed differently. At the ASF these governance specifics are defined at the project level in project bylaws. Most other projects don't make them just as transparent.

Making these specifics transparent though means that projects can evolve independently, avoiding the need to standardise everything across a huge organisation - but makes sure that the differences have to be documented locally.

What do I mean with governance differences in Open Source?

There are projects on GitHub, published purely for the pleasure of the author with no intention for long term maintenance, not intention to fix bugs submitted by users. This would be the equivalent of the governance stage "bug reports and issues welcome" - you can report the bug, but it's on me to find the time to fix it.

There are projects, where the roadmap is created in-house, hidden from public view. Where commit rights come and go with the contract of the employees of one company (think MongoDB, think Elastic, think Tensorflow etc.) - users are welcome to submit patches, they will even be mentored through. All development happens in the open, but control and strategy is never shared. This would be the equivalent of stage "contributions welcome".

There are projects that share write access, but do not share the power to decide who gets write access next. This applies to everyone who is only a committer at an Apache project. This would be the equivalent of "shared write access".

There are projects that are fully shared across multiple independent organisations - think k8s, think any Apache project - those would be shared ownership.


The same levels make sense inside of organisations.

And the same problem as with Open Source happens inside of organisations: "Look, it's on GitHub, so I can submit a PR and it will be open and developed in an open way forever" - a bit like the "if it quaks like a duck, it must a be a duck" - except not all OSS projects are the same ... https://www.slideshare.net/MikeMilinkovich/corporate-shenanigans ... helped me get a glimpse of the depth of games played.

File: patterns/1-initial/governance-levels.md
Comment (MaineC) [inline patterns/1-initial/governance-levels.md]: I like phrasing this as "creating a universally understood language"!

File: patterns/1-initial/governance-levels.md
Comment (MaineC) [inline patterns/1-initial/governance-levels.md:60]: I wonder if we can identify operating models that have evolved in more than one organisation using InnerSource. Given that there are patterns in Open Source I would expect to find patterns in InnerSource orgs as well...

File: patterns/1-initial/governance-levels.md
Comment (MaineC) [inline patterns/1-initial/governance-levels.md]: Great to see an organisation using this pattern internally :)

**Review by MaineC — COMMENTED** [review]
Sorry for the long text :) 

I like the changes, I think the patlet could be a bit more precise - but very obviously I'm not good at capturing my thoughts in short texts ;)

File: patterns/1-initial/governance-levels.md
Comment (spier) [reply patterns/1-initial/governance-levels.md:9]: Some of this text (maybe all of it?) could provide additional value in the `Rationale` section of the pattern.

File: patterns/1-initial/governance-levels.md
Comment (spier) [reply patterns/1-initial/governance-levels.md]: I am not 100% sure but this sounds like agreement to the proposed text change.
Therefore I am resolving this. Feel free to reopen if I misunderstood.

File: patterns/1-initial/governance-levels.md
Comment (spier) [inline patterns/1-initial/governance-levels.md]:
```suggestion
- InnerSource concepts are established within an organisation with multiple projects and teams adopting InnerSource. 
- Internal InnerSource practices are not prescribed in detail.
- Teams/projects have the autonomy to optimise for their local circumstances.
```

Breaking this out into a list, as the checklist format makes it easier for readers to verify if this Context applies to them as well.

File: patterns/1-initial/governance-levels.md
Comment (spier) [inline patterns/1-initial/governance-levels.md]: In terms of "makes the problem difficult", do the "write" and "read" here describe difficulties for two different scenarios or groups of people?

File: patterns/1-initial/governance-levels.md
Comment (spier) [inline patterns/1-initial/governance-levels.md]: What is the motivation of the contributors here?
i.e. why do they "want clarity on the level of influence they can gain"?

File: patterns/1-initial/governance-levels.md
Comment (spier) [reply patterns/1-initial/governance-levels.md:60]: I don't quite follow.

From what I understand, you have identified 3 model below:
- Bug Reports and Issues Welcome
- Contributions Welcome
- Shared Ownership

So are you saying that you would like to know how many other orgs have used these models?

Is this just something that you are interested in in general, or would you like to add information about it to this pattern?

File: patterns/1-initial/governance-levels.md
Comment (spier) [inline patterns/1-initial/governance-levels.md]:
```suggestion
- Using the names within project documentation and contributing guides (see also [Standard Base Documentation](../2-structured/project-setup/base-documentation.md)).
```

linking to relate pattern

File: patterns/1-initial/governance-levels.md
Comment (spier) [inline patterns/1-initial/governance-levels.md]: If you like we could embed the [pyramid](https://innersource.flutter.com/how/pyramid.svg) here to illustrate your solution.

Also would it be useful to mention "Stage 0 · Closed Source" here as well, or do you consider that state not to be an InnerSource model but rather the stage below the models?

Comment (spier) [inline patterns/1-initial/governance-levels.md]:
```suggestion
Flutter Entertainment define an [InnerSource Pyramid](https://innersource.flutter.com/how/) to describe 3 different InnerSource operating models: Readable Source, Guest Contributions and Maintainers in Multiple Teams. Each name is centrally documented. The use of these names is encouraged via repeated usage, direct training and categorisation of each InnerSource project.
```

I am suggesting to use the spelling "InnerSource" here, to stay consistent when compared to other patterns. Even though I know that Flutter spells this differently.
Also we can leave the quotation marks away, as it is clear through the link that this is a unique term.

Further, should we use the exact term as described in the Solution here, i.e. "InnerSource operating models"?

File: patterns/1-initial/governance-levels.md
Comment (spier) [inline patterns/1-initial/governance-levels.md:3]: For a discussion about alternative names for this pattern see
https://github.com/InnerSourceCommons/InnerSourcePatterns/pull/357#issuecomment-939138250

Btw I suggest that we work out all other content of this pattern first, and the handle the name as the very last thing. Naming is so hard! :)

File: patterns/1-initial/governance-levels.md
Comment (spier) [inline patterns/1-initial/governance-levels.md]:
```suggestion
Structured
```

With these great improvements we should absolutely move this pattern to the next level. 🚀

Comment (spier) [reply patterns/1-initial/governance-levels.md]: FYI I suggest that we move the pattern to the correct folder and place in the README in a subsequent PR. I can handle that once we get there.

File: patterns/1-initial/governance-levels.md
Comment (spier) [inline patterns/1-initial/governance-levels.md:84]: @MaineC I know we talked about it but I forgot (and it is probably best to have have these conversations in the open anyways):

Are you already using this pattern at Europace? 
If yes, do you want to list your org here, with 2-3 sentences describing how you are using the pattern?

**Review by spier — CHANGES_REQUESTED** [review]
Great work on this PR @robtuley!

I left some comments, including some questions for @MaineC. 

Btw: I am sorry that it took me so long to get around to review this!

File: patterns/1-initial/governance-levels.md
Comment (spier) [inline patterns/1-initial/governance-levels.md]:
```suggestion
- InnerSource concepts are established within an organisation with multiple projects and teams adopting InnerSource.
```

Comment (spier) [reply patterns/1-initial/governance-levels.md]: Sorry for making the linter unhappy there :)

File: patterns/1-initial/governance-levels.md
Comment (robtuley) [reply patterns/1-initial/governance-levels.md]: I was simply thinking of one set of people producing some docs (write) and then the wider community consumption effort (read). But the question makes me think this is unnecessary complexity so have updated to remove `to write and to read`.

File: patterns/1-initial/governance-levels.md
Comment (robtuley) [reply patterns/1-initial/governance-levels.md]: Good point. Changed to:

> Users want clarity on the level of influence they can gain in an InnerSource project when deciding whether or how to use it.

In my experience people are most keen on understanding what is possible with the right effort when they are making a decision on whether to use a project, or ramp up their usage or similar. In effect they are working out what they are getting into -- i.e. is this a dependency they might be blocked by or is contribution on a sensible timescale realistic.

File: patterns/1-initial/governance-levels.md
Comment (robtuley) [reply patterns/1-initial/governance-levels.md]: We consider the 'stage 0' state not to be InnerSource, although we'd doing some work at the moment to try and expand how we express this internally to better cover other sharing models like delegating something to a team with closed source. 

Added the pyramid svg image: is the format right I copied existing pattern images `Flutter InnerSource Pyramid` ..?

File: patterns/1-initial/governance-levels.md
Comment (robtuley) [reply patterns/1-initial/governance-levels.md:9]: This is some insightful comparisons to opensource so it would be good to incorporate somehow, will take a crack at that later this week. I do stand by the "... the term is not precise enough to ..." language though because I don't think it misses the point: InnerSource is simply not a precise term and therefore needs additional qualifiers when discussing at anything other than a high level.

File: patterns/1-initial/governance-levels.md
Comment (spier) [inline patterns/1-initial/governance-levels.md]: Note: We need to check whether SVG gets displayed correctly when publishing to our book (on gitbook.com). If not, then we can use PNG instead.

File: patterns/1-initial/governance-levels.md
Comment (spier) [inline patterns/1-initial/governance-levels.md]: What is meant by "sub-options" here?
i.e. sub below what?

File: patterns/1-initial/governance-levels.md
Comment (spier) [inline patterns/1-initial/governance-levels.md]: Are the two paragraphs here two different stories, both illustrating the same problem?
Or is it meant to be read as a single story?

If they are two distinct stories, then we might want to indicate that somehow.

File: patterns/1-initial/governance-levels.md
Comment (spier) [inline patterns/1-initial/governance-levels.md]:
```suggestion
The solution is to create a universally understood language to describe the InnerSource operating models that are used in your organisation. 

We define **InnerSource operating model** as a description of how much influence the core development team of a project ist willing to share with contributing teams. Or in other terms, the level of influence a contributing team can gain in the respective project.

The shared language for these InnerSource operating models can be established with these steps:
```


The specific building blocks that this pattern defines are the InnerSource operating models.

Therefore I am wondering if we should leave the phrase "building block" away, as that suggests that there may be even further building blocks that we want to define. Instead I would define here what we mean by "InnerSource operating models".

I tried to do that above, in parts be reusing some of the text from the left.

File: patterns/1-initial/governance-levels.md
Comment (spier) [inline patterns/1-initial/governance-levels.md]: The Solution is using variations of terms when talking about the "model":
- InnerSource operating model
- model
- name

I believe that "name" is used to emphasize that we are definition a specific language here?

When reading the pattern from top to bottom, at this point I hd already forgotten what "name" was referring to. So maybe we could make it easier for the reader by calling it "model names" here, to repeat what names we are referring to?

```suggestion
Examples of promoting the model names (3) are:
```

File: patterns/1-initial/governance-levels.md
Comment (spier) [reply patterns/1-initial/governance-levels.md:84]: This comment here shouldn't block this PR. We can always add more Known Instances later.

**Review by spier — COMMENTED** [review]
I gave this a 2nd reading and left some comments.

File: patterns/1-initial/governance-levels.md
Comment (spier) [reply patterns/1-initial/governance-levels.md:60]: When reading this again, I think I understand it now.
I think what @MaineC meant here was whether the examples of the operating models that we are listing below are indeed common in the industry.

I am claiming that we will find this out as this pattern matures.
i.e. every time we find a new Known Instances, we can ask the respective org which models they use. That way we should find the common ones.

So for the moment I suggest that there is nothing left to do here, and I am closing this thread.

File: patterns/1-initial/governance-levels.md
Comment (spier) [reply patterns/1-initial/governance-levels.md]: Note to self: We will also need markdown notation for the image, in order for gitbook to pick it up correctly.

File: patterns/1-initial/governance-levels.md
Comment (robtuley) [reply patterns/1-initial/governance-levels.md]: sub to "innersource", but have removed 'sub' if it is confusing.

File: patterns/1-initial/governance-levels.md
Comment (spier) [reply patterns/1-initial/governance-levels.md]:
```suggestion
![InnerSource Pyramid used by Flutter Entertainment](../../assets/img/flutter-pyramid.svg)
```

File: patterns/1-initial/governance-levels.md
Comment (robtuley) [reply patterns/1-initial/governance-levels.md:9]: I wanted to capture your comment here for posterity, so incorporated it into this PR in 5095397

File: patterns/1-initial/governance-levels.md
Comment (robtuley) [reply patterns/1-initial/governance-levels.md]: 2 examples, so I separated with sub-headings. I also incorporated [this fab comment](https://github.com/InnerSourceCommons/InnerSourcePatterns/pull/357#discussion_r753125084) under an additional sub-heading here.

**Comment by robtuley** [comment]
Right -- revisited this to get it over the line. Think I've covered all outstanding review comments but quite the discussion here so by all means point me to any unresolved points I've missed. I propose to merge as-is, and then follow up with a new PR(s) to:

- tackle the pattern name ([ref](https://github.com/InnerSourceCommons/InnerSourcePatterns/pull/357#discussion_r790251361))
- chase further known instances
- move to structured pattern to publish to book
- tackle any svg problems (or convert to png)

File: patterns/1-initial/governance-levels.md
Comment (spier) [reply patterns/1-initial/governance-levels.md:84]: No reply. Resolving.

**Review by spier — APPROVED** [review]
I like the approach of "merge as-is, and then follow up with a new PR(s)".
Great stuff!

**Comment by MaineC** [comment]
I really like how this pattern now references the equivalent in the Open Source world - it even comes with specific examples (though likely MongoDB and Elastic should be listed as "formerly Open Source", as both switched license to a proprietary one a long while ago). Tensorflow is a great example, as it is Open Source, but also according to whisper networks said to slow down or reject contributions that only benefit the competitors of the employer of it's current team of maintainers.

One question: What is the reason for using the term "Maintainer" in this document instead of the more common InnerSource term "Trusted Committer"?

**Comment by robtuley** [comment]
> One question: What is the reason for using the term "Maintainer" in this document instead of the more common InnerSource term "Trusted Committer"?

I think the only usage of "Maintainer" in this doc is the reference at the end to our internal Flutter pyramid names. Internally here we use the term "Maintainer" rather than "Trusted Committer" simply for historical reasons -- it was used for a number of years and became part of our common language internally so there seemed no value to change it. In fact we used to use the term "shared development" to refer to InnerSource, but made an explicit effort to change this internally to make wider information easier to find.

**Pull request merged**
